procinfo/src/macos.rs

```diff
@@ -227,3 +198,97 @@ impl LocalProcessInfo {
         }
     }
 }
+
+fn parse_exe_and_argv_sysctl(buf: Vec<u8>) -> Option<(PathBuf, Vec<String>)> {
+    use libc::c_int;
+
+    // The data in our buffer is laid out like this:
+    // argc - c_int
+    // exe_path - NUL terminated string
+    // argv[0] - NUL terminated string
+    // argv[1] - NUL terminated string
+    // ...
+    // argv[n] - NUL terminated string
+    // envp[0] - NUL terminated string
+    // ...
+
+    let mut ptr = &buf[0..buf.len()];
+
+    let argc: c_int = unsafe { std::ptr::read(ptr.as_ptr() as *const c_int) };
+    ptr = &ptr[std::mem::size_of::<c_int>()..];
+
+    fn consume_cstr(ptr: &mut &[u8]) -> Option<String> {
+        // Trim out trailing nulls
+        let not_nul = ptr.iter().position(|&c| c != 0)?;
+        *ptr = ptr.get(not_nul..)?;
+
+        // Then parse to the end of the null terminated string
+        let nul = ptr.iter().position(|&c| c == 0)?;
```

Comment:
From a safety perspective, the biggest concern with the pointer manipulation is whether we walk outside of the buffer.  I think this code is OK because of the way that the slice manipulations are written, but it would be good to add a test case where we have multiple trailing zeroes at the end of the buffer to assert that we don't somehow walk off the very end

Reply:
also one where there are no trailing zeroes!